File: dynamic programing imp questions/sortstack.cpp
#include<bits/stdc++.h>
using namespace std;


void insert(stack<int>&st,int x)
{
	
	
	if(st.size()==0||x<=st.top())  // x>=st.top() for descendiung order
	{
		st.push(x);
		return;
	}
	
	int te=st.top();
	st.pop();
	insert(st,x);
	st.push(te);
}

void sortstack(stack<int>&st)
{
	
	
	if(st.size()==1)
	{
		return;
	}
	
	int temp=st.top();
	
	st.pop();
	
	sortstack(st);
	
	insert(st,temp);
	
}

int main()
{
	
	
	stack<int>s;
	
	s.push(1);
	s.push(5);
	s.push(3);
	s.push(0);
	s.push(-1);
	s.push(9);
	
	
	stack<int>temp=s;
	
	while(temp.empty()==false)
	{
		
		cout<<temp.top();
		temp.pop();
	}
	
	
	sortstack(s);
	
	
	cout<<endl;
	
	while(!s.empty())
	{
		cout<<s.top()<<" ";
		s.pop();
	}
	
}

File: Big-Java-codes/GoldManPractice/src/SortStack.java
import java.util.Stack;

public class SortStack {

	
	public static void insert(Stack<Integer>st,int x)
	{
		
		if(st.size()==0||x>=st.peek())
		{
			st.push(x);
		}
		int tem=st.pop();
		//st.pop();
		insert(st,x);
		st.push(tem);
	}
	
	public static void sortStack(Stack<Integer>st)
	{
		
		if(st.size()==1)
			return;
		
		int temp=st.pop();
		//st.pop();
		sortStack(st);
		insert(st,temp);
		
		return;
		
	}
	
	public static void main(String[] args)
	{
		
		Stack<Integer>st=new Stack<>();
		
		st.push(5);
		st.push(1);
		st.push(0);
		st.push(7);
		st.push(2);
		st.push(6);
		
		System.out.println(st);
		
		
		
		sortStack(st);
		
		
		System.out.println(st);
		
		
	}
}
